OpenRA.Game/Actor.cs

```diff
@@ -265,6 +265,11 @@ public void AddTrait(object trait)
 
 		public void Dispose()
 		{
+			// If CurrentActivity isn't null, run OnActorDispose in case some cleanups are needed.
+			// This should be done before the FrameEndTask to avoid dependency issues.
+			if (CurrentActivity != null)
+				CurrentActivity.RootActivity.OnActorDispose(this);
```

Comment:
Do we need to worry about propagating this to child activities, or do we assume that the activities will handle this themselves if they use them? My gut feeling is that this probably should be automatic.

Reply:
I lean towards agreeing that it should be automatic, that would require changing the protection level of ChildActivity, though.
Personally I wouldn't really mind, just wanted to point it out in case there are some potential issues with that I'm not aware of.

Reply:
One option is to define a `OnActorDisposeOuter` on `Activity` (it should have `internal` visibility so that mod code can't see it), which calls `OnActorDisposeOuter` on the child activity if it exists and then calls `OnActorDispose` on itself.  We use this pattern in other places, notably widgets.  We may also want to propagate this through the `NextActivity` chain too?

If this isn't going to work then I think a reasonable compromise is to add a comment to the default `OnActorDispose` definition explaining that it must be manually chained through by any activities that use child activities.

Reply:
Appears to work fine, so I'd like to go with it.

Not sure what you mean with propagating through the NextActivity chain, though?

Reply:
I suppose calling `OnActorDisposeOuter` on `NextActivity` (and that activity will call it on its `NextActivity` and so on, hence "through the chain").

Reply:
But do we really want or need that?

When the actor is disposed, NextActivity hasn't even started yet, or it wouldn't be "Next" anymore. I don't see why we would want to run the dispose-specific code on activities that haven't even started yet, since that would mean OnFirstRun hasn't run, either. I'm not even sure if the ctor is run before the activity becomes 'current'.

Reply:
Leaving changes to NextActivity for the future and opened #15632 instead, as discussed on IRC.